### PR TITLE
Added experimental multiresolution STFT loss using auraloss.

### PR DESCRIPTION
**Pull request by honkkis**
Adding experimental multiresolution STFT loss from auraloss package in addition to the normal loss, antialias seems to be reduced based on the experiments here: https://github.com/honkkis/neural-amp-modeler/blob/WIP_stft_loss/dev/stft/stft_tests1.png , where top is the target, middle is with the MRSTFT loss, and bottom is the normal loss.

### CODE REVIEW

**Comment by honkkis** [comment]
Training with this loss added does not seem to make other losses worse, but it has quite a big effect to the spectral loss https://github.com/honkkis/neural-amp-modeler/blob/WIP_stft_loss/dev/stft/mr_stft_losses.png

File: nam/models/base.py [hunk 1]
```diff
@@ -259,3 +275,20 @@ def _mse_loss(self, preds, targets, pre_emph_coef: Optional[float] = None):
                 z[..., 1:] - pre_emph_coef * z[..., :-1] for z in (preds, targets)
             ]
         return nn.MSELoss()(preds, targets)
+
+    def _mrstft_loss(self, preds: torch.Tensor, targets: torch.Tensor) -> torch.Tensor:
```
Comment (sdatkinson) [inline nam/models/base.py:279]: Is there a reason that this has to be a method of  `Model`? If not, i'd rather see it implemented in `losses.py` where the others are.

File: nam/models/base.py [hunk 2]
```diff
@@ -259,3 +275,20 @@ def _mse_loss(self, preds, targets, pre_emph_coef: Optional[float] = None):
...
+        """
+        Experimental Multi Resolution Short Time Fourier Transform Loss using auraloss implementation.
+        B: Batch size
+        L: Sequence length
+
+        :param preds: (B,L)
+        :param targets: (B,L)
+        :return: ()
+        """
+        device = 'cpu' # not all platforms support this on gpu yet
```
Comment (sdatkinson) [inline nam/models/base.py:289]: > `# not all platforms support this on gpu yet`

Ooh really? Is there a custom kernel that's needed?

What's the performance hit (i.e. iteration speed) look like for having to do this? (I assume not the end of the world and I don't think I'd reject the PR over it; just curious)

File: nam/models/base.py [hunk 3]
```diff
@@ -173,6 +175,15 @@ def parse_config(cls, config):
     def net(self) -> nn.Module:
         return self._net
 
+    def initialize_losses(self):
+        if self._loss_config.mstft_weight > 0.0:
+            import auraloss
+            self._mrstft = auraloss.freq.MultiResolutionSTFTLoss()
```
Comment (sdatkinson) [inline nam/models/base.py:181]: What's the reason for doing this instead of 
```python
from .losses import mrstft_loss

# then e.g. line 219
loss = loss + self._loss_config.mstft_weight * mrstft_loss(preds, targets)
```
?

Seems like we could cut out lines 178-186 and not need to define `self._mrstft`.

If there's not a clear advantage, then I'd prefer what I've written here so that it's consistent with the other losses' implementations.

**Review by sdatkinson — COMMENTED** [review]
Thanks for the PR! Looks really exciting.

For approval, please either move the loss to `losses.py` as the comments describe or let me know the justification for why it's implemented as-is.

Thanks again!

File: nam/models/base.py [hunk 4]
```diff
@@ -54,6 +54,7 @@ class LossConfig(InitializableFromConfig):
         https://www.mdpi.com/2076-3417/10/3/766. Paper value: 0.95.
     """
 
+    mstft_weight: float = 0.0 # 0.0 means no multiresolution stft loss, 2e-4 works pretty well if one wants to use it
```
[the remark this answers is not in the document]
Comment (honkkis) [reply nam/models/base.py:57]: Thanks :)

File: nam/models/base.py (hunk 1 above)
Comment (honkkis) [reply nam/models/base.py:279]: Sure, but we need to instantiate the MultiResolutionSTFTLoss and store it somewhere - I think instantiating it in every loss function call is not a good idea. All the previous losses are stateless and therefore this needs different handling. https://github.com/csteinmetz1/auraloss#usage

Another alternative would be to have AuraLoss class defined in losses.py and its instance would be stored in Model.

File: nam/models/base.py (hunk 2 above)
Comment (honkkis) [reply nam/models/base.py:289]: Did not work in apple metal GPU (uses unimplemented methods), have not tried yet in nvidia.

File: nam/models/base.py (hunk 3 above)
Comment (honkkis) [reply nam/models/base.py:181]: Sure, but we need to instantiate the MultiResolutionSTFTLoss and store it somewhere - I think instantiating it in every loss function call is not a good idea. All the previous losses are stateless and therefore this needs different handling. https://github.com/csteinmetz1/auraloss#usage

Another alternative would be to have AuraLoss class defined in losses.py and its instance would be stored in Model.

File: nam/models/base.py (hunk 1 above)
Comment (sdatkinson) [reply nam/models/base.py:279]: Hmm. Interesting.

On measure, I think I'd rather have this and I'm happy to explore it more deeply on my own time :)

**Comment by sdatkinson** [comment]
@cdicle that's a good point.

But, I'm happy to take this PR as-is, enjoy the benefits, and iterate on the details separately.

**Review by sdatkinson — APPROVED** [review]
LGTM! 👍🏻

**Comment by cdicle** [comment]
@sdatkinson @honkkis auraloss dependency missing.

**Comment by sdatkinson** [comment]
Ugh, I could have sworn I saw those file diffs.

I'll grab it when I get home.

(Note to self: no PR approvals w/o test coverage)

**Comment by sdatkinson** [comment]
Got it, #153

**Comment by honkkis** [comment]
Thanks, I forgot to add the dependency, thanks for adding it also!
